### PR TITLE
accounts/abi: distinguish 'Unpack' error

### PR DESCRIPTION
**Pull request by levisyin**
In my opinion, if we need to return different errors for different parameter check, we should split them, otherwise we should put them together 
```go
if len(data) < 4 || !bytes.Equal(data[:4], e.ID[:4]) {
	return "", errors.New("invalid data for unpacking")
}
```

### CODE REVIEW

File: accounts/abi/error.go [hunk 1]
```diff
@@ -84,10 +84,10 @@ func (e Error) String() string {
 
 func (e *Error) Unpack(data []byte) (interface{}, error) {
 	if len(data) < 4 {
-		return "", errors.New("invalid data for unpacking")
+		return "", errors.New("invalid data for unpacking, length is less than 4")
```
Comment (MariusVanDerWijden) [inline accounts/abi/error.go:87]: Maybe something like this?
```suggestion
		return "", fmt.Errorf("invalid data length for unpacking: got %v, want 4", len(data))
```

File: accounts/abi/error.go [hunk 2]
```diff
@@ -84,10 +84,10 @@ func (e Error) String() string {
...
 	}
 	if !bytes.Equal(data[:4], e.ID[:4]) {
-		return "", errors.New("invalid data for unpacking")
+		return "", errors.New("invalid data for unpacking, first four bytes are not equal with error's signature")
```
Comment (MariusVanDerWijden) [inline accounts/abi/error.go:90]:
```suggestion
		return "", fmt.Errorf("invalid event ID, got %v want %v", data[:4], e.ID[:4])
```

**Review by MariusVanDerWijden — COMMENTED** [review]
Good catch, these errors should be distinct. Would be nice to get some more context on them

**Review by MariusVanDerWijden — APPROVED** [review]
lgtm

File: accounts/abi/error.go [hunk 3]
```diff
@@ -84,10 +83,10 @@ func (e Error) String() string {
 
 func (e *Error) Unpack(data []byte) (interface{}, error) {
 	if len(data) < 4 {
-		return "", errors.New("invalid data for unpacking")
+		return "", fmt.Errorf("invalid data length for unpacking: got %v, want 4", len(data))
```
Comment (holiman) [inline accounts/abi/error.go:86]:
```suggestion
		return "", fmt.Errorf("insufficient data for unpacking: have %d, want at least 4", len(data))
```

File: accounts/abi/error.go [hunk 4]
```diff
@@ -84,10 +83,10 @@ func (e Error) String() string {
...
 	}
 	if !bytes.Equal(data[:4], e.ID[:4]) {
-		return "", errors.New("invalid data for unpacking")
+		return "", fmt.Errorf("invalid event ID, got %v want %v", data[:4], e.ID[:4])
```
Comment (holiman) [inline accounts/abi/error.go:89]:
```suggestion
		return "", fmt.Errorf("invalid identifier, have %#x want %#x", data[:4], e.ID[:4])
```

**Review by holiman — COMMENTED** [review]
Minor nits

File: accounts/abi/error.go (hunk 4 above)
Comment (levisyin) [reply accounts/abi/error.go:89]: Done

File: accounts/abi/error.go (hunk 3 above)
Comment (levisyin) [reply accounts/abi/error.go]: Done

**Review by holiman — APPROVED** [review]
LGTM

**Pull request merged**